### DISCUSSION TITLE
Plugin search

### DISCUSSION THREAD

**Discussion by finestructure** [SwiftPackageIndex/SwiftPackageIndex-Server#1661] [Ideas]
Joe @heckj recently suggested we add support for SPM plugins to search and I wanted to capture this here.

His idea is that we parse plugin capabilities from packages and that there'd be a way to search for them in the index.

It should require the following:

- [x] add support for Swift 5.6
- [x] extend product type parsing to handle plugins
- [x] extend search materialised view to contain product types
- [x] extend search filter to allow searching for product types

We've added Swift 5.6 support this week, which was one of the bigger hurdles in this.

Plugins expose `.plugin` products and it should be quite simple to ingest them alongside the `executable` and `library` products we're already reading.

![CleanShot 2022-03-15 at 08 15 28@2x](https://user-images.githubusercontent.com/65520/161058480-0fffdd88-913f-49d3-b6d7-2cc301391a01.png)

Finally, we'd need to add a new search filter to search for those products. The way I envision this would be a new `product` filter, which could/should handle `executable` and `library` as well as `plugin`:

```
product:plugin
product:!executable
```

**Comment by daveverwer** [comment]
Just to add one more item to the list above:

- [x] Add "X plugins" next to where we give library and executable information.

<img width="558" alt="Screenshot 2022-03-31 at 14 01 12@2x" src="https://user-images.githubusercontent.com/5180/161060737-d5f0e5c7-5065-496e-8666-864cfa165377.png">

**Comment by icanzilb** [comment]
@finestructure I just found that there is already a thread about searching by product type here. Looking at your initial writeup, what do you think is the use case for using a negative filter? In other words, what's the use case for searching for packages that do not offer a "plugin" or do not offer an "executable" product?

**Reply by finestructure** [reply]
I think I mainly listed that for sake of completeness, TBH. Simply being able to filter on a certain type should be all we need really. Although I suspect we might get the `!` negation pretty much for free IIRC how the filters work, it's probably not really all that helpful!

**Reply by icanzilb** [reply]
I'm just saying, if the feature is available it needs to be tested and maintained. If we can't think of a use-case maybe we can skip it 🤷🏽‍♂️

**Reply by finestructure** [reply]
Yea, agreed!

**Comment by icanzilb** [comment]
Ok, I think at this point https://github.com/SwiftPackageIndex/SwiftPackageIndex-Server/pull/1810 addresses everything  saw mentioned in this thread. I think this could be an amazing feature to aid discoverability of plugins
